### PR TITLE
Globally Unique FATE Transaction Ids - Part 3

### PR DESCRIPTION
**Pull request by kevinrr888**
This addresses several previously deferred changes for issue #4044. Root of most of these new changes were from changing TabletMetadata and TabletUpdates (in Ample) - FateId is now stored in the Metadata table instead of just the formatted long id. Addresses deferred changes to TabletMetadata, TabletUpdates, CompactionConfigStorage, SelectedFiles, and Ample.

### CODE REVIEW

File: core/src/main/java/org/apache/accumulo/core/metadata/schema/CompactionMetadata.java [hunk 1]
```diff
@@ -98,7 +99,7 @@ private static class GSonData {
     String groupId;
     short priority;
     boolean propDels;
-    Long fateTxId;
```
Comment (keith-turner) [inline core/src/main/java/org/apache/accumulo/core/metadata/schema/CompactionMetadata.java:101]: We may need a follow on update for upgrade

File: core/src/main/java/org/apache/accumulo/core/metadata/schema/CompactionMetadata.java [hunk 2]
```diff
@@ -111,7 +112,7 @@ public String toJson() {
     jData.groupId = cgid.toString();
     jData.priority = priority;
     jData.propDels = propagateDeletes;
-    jData.fateTxId = fateTxId;
+    jData.fateId = fateId;
```
Comment (keith-turner) [inline core/src/main/java/org/apache/accumulo/core/metadata/schema/CompactionMetadata.java:115]: The following would probably make mch more readable json and be tolerant of changes to the implementation of FateId like variable renames.

```suggestion
    jData.fateId = fateId.canonical();
```

File: core/src/main/java/org/apache/accumulo/core/metadata/schema/CompactionMetadata.java [hunk 3]
```diff
@@ -121,7 +122,7 @@ public static CompactionMetadata fromJson(String json) {
     return new CompactionMetadata(jData.inputs.stream().map(StoredTabletFile::new).collect(toSet()),
         StoredTabletFile.of(jData.tmp).getTabletFile(), jData.compactor,
         CompactionKind.valueOf(jData.kind), jData.priority,
-        CompactorGroupIdImpl.groupId(jData.groupId), jData.propDels, jData.fateTxId);
+        CompactorGroupIdImpl.groupId(jData.groupId), jData.propDels, jData.fateId);
```
Comment (keith-turner) [inline core/src/main/java/org/apache/accumulo/core/metadata/schema/CompactionMetadata.java:125]: This suggestion goes with another comment.

```suggestion
        CompactorGroupIdImpl.groupId(jData.groupId), jData.propDels, FateId.from(jData.fateId));
```

File: core/src/main/java/org/apache/accumulo/core/metadata/schema/MetadataSchema.java [hunk 4]
```diff
@@ -334,18 +334,13 @@ public static class BulkFileColumnFamily {
...
-      public static long getBulkLoadTid(String vs) {
-        if (FateTxId.isFormatedTid(vs)) {
-          return FateTxId.fromString(vs);
-        } else {
-          // a new serialization format was introduce in 2.0. This code support deserializing the
-          // old format.
-          return Long.parseLong(vs);
-        }
+      public static FateId getBulkLoadTid(String vs) {
+        // ELASTICITY_TODO issue 4044 will this be an issue removing the old code?
```
Comment (keith-turner) [inline core/src/main/java/org/apache/accumulo/core/metadata/schema/MetadataSchema.java:342]:
```suggestion
        // ELASTICITY_TODO issue 4044 will this be an issue removing the old code?  May need to introduce code in upgrade to handle old format.
```

File: core/src/main/java/org/apache/accumulo/core/metadata/schema/SelectedFiles.java [hunk 5]
```diff
@@ -43,26 +43,26 @@ public class SelectedFiles {
...
+    this.fateId = fateId;
   }
 
   private static class SelectedFilesTypeAdapter extends TypeAdapter<SelectedFiles> {
 
     @Override
     public void write(JsonWriter out, SelectedFiles selectedFiles) throws IOException {
       out.beginObject();
-      out.name("txid").value(FateTxId.formatTid(selectedFiles.getFateTxId()));
+      out.name("txid").value(selectedFiles.getFateId().canonical());
```
Comment (keith-turner) [inline core/src/main/java/org/apache/accumulo/core/metadata/schema/SelectedFiles.java:65]:
```suggestion
      out.name("fateId").value(selectedFiles.getFateId().canonical());
```

File: core/src/main/java/org/apache/accumulo/core/metadata/schema/SelectedFiles.java [hunk 6]
```diff
@@ -90,7 +90,7 @@ public SelectedFiles read(JsonReader in) throws IOException {
         String name = in.nextName();
         switch (name) {
           case "txid":
```
Comment (keith-turner) [inline core/src/main/java/org/apache/accumulo/core/metadata/schema/SelectedFiles.java:92]:
```suggestion
          case "fateId":
```

File: server/base/src/main/java/org/apache/accumulo/server/compaction/CompactionConfigStorage.java [hunk 7]
```diff
@@ -83,29 +85,30 @@ public static CompactionConfig getConfig(ServerContext context, long fateTxId,
...
 
-    Map<Long,CompactionConfig> configs = new HashMap<>();
+    Map<FateId,CompactionConfig> configs = new HashMap<>();
 
     var children = context.getZooReaderWriter()
         .getChildren(context.getZooKeeperRoot() + Constants.ZCOMPACTIONS);
     for (var child : children) {
-      var fateTxid = Long.parseLong(child, 16);
-      var cconf = getConfig(context, fateTxid, tableIdPredicate);
+      String[] fields = child.split(DELIMITER);
```
Comment (keith-turner) [inline server/base/src/main/java/org/apache/accumulo/server/compaction/CompactionConfigStorage.java:107]:
```suggestion
      String[] fields = child.split(DELIMITER);
      Preconditions.checkState(fields.length == 2, "Unexpected child %s", child);
```

File: server/base/src/main/java/org/apache/accumulo/server/manager/state/TabletManagementParameters.java [hunk 8]
```diff
@@ -205,7 +206,7 @@ private static class JsonData {
 
     final Map<String,Set<String>> tserverGroups;
 
-    final Map<Long,Map<String,String>> compactionHints;
+    final Map<FateId,Map<String,String>> compactionHints;
```
Comment (keith-turner) [inline server/base/src/main/java/org/apache/accumulo/server/manager/state/TabletManagementParameters.java:209]: Making the key type for this map a string that is the canonical value would yield nicer and more stable json.

File: test/src/main/java/org/apache/accumulo/test/functional/AmpleConditionalWriterIT.java [hunk 9]
```diff
@@ -745,58 +751,70 @@ public void testCompacted() {
 
       var ctmi = new ConditionalTabletsMutatorImpl(context);
 
+      FateInstanceType type = FateInstanceType.fromTableId(tid);
+      FateId fateId55L = FateId.from(type, 55L);
+      FateId fateId45L = FateId.from(type, 45L);
+      FateId fateId56L = FateId.from(type, 56L);
+      FateId fateId65L = FateId.from(type, 65L);
+      FateId fateId75L = FateId.from(type, 75L);
```
Comment (keith-turner) [inline test/src/main/java/org/apache/accumulo/test/functional/AmpleConditionalWriterIT.java:759]: Reordered the vars to be sorted.

```suggestion
      FateId fateId45L = FateId.from(type, 45L);
      FateId fateId55L = FateId.from(type, 55L);
      FateId fateId56L = FateId.from(type, 56L);
      FateId fateId65L = FateId.from(type, 65L);
      FateId fateId75L = FateId.from(type, 75L);
```

File: test/src/main/java/org/apache/accumulo/test/functional/SplitRecoveryIT.java [hunk 10]
```diff
@@ -46,6 +46,8 @@
 import org.apache.accumulo.core.data.TableId;
 import org.apache.accumulo.core.data.Value;
 import org.apache.accumulo.core.dataImpl.KeyExtent;
+import org.apache.accumulo.core.fate.FateId;
+import org.apache.accumulo.core.fate.FateInstanceType;
```
Comment (keith-turner) [inline test/src/main/java/org/apache/accumulo/test/functional/SplitRecoveryIT.java:50]: #4260 will conflict with these changes,  will hold off on merging it until this PR is merged.

File: server/manager/src/main/java/org/apache/accumulo/manager/tableOps/bulkVer2/CleanUpBulkImport.java [hunk 11]
```diff
@@ -108,13 +108,12 @@ private static void removeBulkLoadEntries(Ample ample, TableId tableId, FateId f
...
-          // ELASTICITY_TODO DEFERRED - ISSUE 4044
-          if (tablet.getLoaded().values().stream().anyMatch(l -> l == fateId.getTid())) {
+          if (tablet.getLoaded().values().stream()
+              .anyMatch(loadedFateId -> loadedFateId.equals(fateId))) {
             var tabletMutator =
                 tabletsMutator.mutateTablet(tablet.getExtent()).requireAbsentOperation();
-            tablet.getLoaded().entrySet().stream()
-                .filter(entry -> entry.getValue() == fateId.getTid()).map(Map.Entry::getKey)
-                .forEach(tabletMutator::deleteBulkFile);
+            tablet.getLoaded().entrySet().stream().filter(entry -> entry.getValue().equals(fateId))
```
Comment (keith-turner) [inline server/manager/src/main/java/org/apache/accumulo/manager/tableOps/bulkVer2/CleanUpBulkImport.java:115]: Hard to tell from looking at this diff, but I assume the types line up?  Meaning `entry.getValue()` is a FateId type.

File: server/base/src/main/java/org/apache/accumulo/server/manager/state/TabletManagementParameters.java (hunk 8 above)
Comment (keith-turner) [reply server/base/src/main/java/org/apache/accumulo/server/manager/state/TabletManagementParameters.java:209]: If this seems like a big change, could add an ELASTICITY_TODO for this to get this PR merged.

File: server/manager/src/main/java/org/apache/accumulo/manager/tableOps/bulkVer2/CleanUpBulkImport.java (hunk 11 above)
Comment (kevinrr888) [reply server/manager/src/main/java/org/apache/accumulo/manager/tableOps/bulkVer2/CleanUpBulkImport.java:115]: Yes, types line up. TabletMetadata.getLoaded() (tablet.getLoaded()) returns Map<StoredTabletFile, FateId>

Comment (keith-turner) [reply server/manager/src/main/java/org/apache/accumulo/manager/tableOps/bulkVer2/CleanUpBulkImport.java:115]: Thanks

**Comment by kevinrr888** [comment]
Completed review changes:
- FateId refactored (better Pattern for FateId and simpler validation)
- CompactionMetadata GSonData changed to use the canonical FateId string instead of FateId (note that old data may need to be considered here)
- SelectedFiles renamed "txid" --> "fateId". SelectedFilesTest updated to be consistent with changes
- CompactionConfigStorage added Preconditions check
- Changed compactionHints in TabletManagementParameters from Map<FateId,Map<String,String>> --> Map<String,Map<String,String>>: required changes to TabletManagementParameters, CompactionConfigStorage, CompactionJobGenerator, TabletManagementParametersTest, Manager, CancelCompactions, and PreDeleteTable
- AmpleConditionalWriterIT.testCompacted() reordering of FateIds

**Comment by keith-turner** [comment]
> Changed compactionHints in TabletManagementParameters from Map<FateId,Map<String,String>> --> Map<String,Map<String,String>>: required changes to TabletManagementParameters, CompactionConfigStorage, CompactionJobGenerator, TabletManagementParametersTest, Manager, CancelCompactions, and PreDeleteTable

Sorry I did not communicate very clearly for the comment about that.  Was not looking to change the external map type parameters for TabletManagementParameters to have a key type of String.  Just thought it would be good to change internal map used for json so that the json encoding would cleaner.  So was thinking  of pattern like the following.

 * The public methods on TabletManagementParameters use `Map<FateId,Map<String,String>>`
 * Internally for json encoding `Map<FateId,Map<String,String>>` is converted to/from `Map<String,Map<String,String>>`

**Comment by kevinrr888** [comment]
> > Changed compactionHints in TabletManagementParameters from Map<FateId,Map<String,String>> --> Map<String,Map<String,String>>: required changes to TabletManagementParameters, CompactionConfigStorage, CompactionJobGenerator, TabletManagementParametersTest, Manager, CancelCompactions, and PreDeleteTable
> 
> Sorry I did not communicate very clearly for the comment about that. Was not looking to change the external map type parameters for TabletManagementParameters to have a key type of String. Just thought it would be good to change internal map used for json so that the json encoding would cleaner. So was thinking of pattern like the following.
> 
> * The public methods on TabletManagementParameters use `Map<FateId,Map<String,String>>`
> * Internally for json encoding `Map<FateId,Map<String,String>>` is converted to/from `Map<String,Map<String,String>>`

Ah gotcha. Will fix.

**Review by keith-turner — APPROVED** [review]

**Comment by dlmarion** [comment]
@kevinrr888 - The latest IT run is failing (stacktrace below). I think it might be due to this merge. Just an FYI...

```
org.opentest4j.AssertionFailedError: Test json should be identical to actual metadata at this point ==> expected: <{"fateId":"FATE:USER:0000000000000002","selAll":true,"files":["{\"path\":\"hdfs://localhost:8020/accumulo/tables/2a/default_tablet/F0000070.rf\",\"startRow\":\"\",\"endRow\":\"\"}","{\"path\":\"hdfs://localhost:8020/accumulo/tables/2a/default_tablet/F0000071.rf\",\"startRow\":\"\",\"endRow\":\"\"}","{\"path\":\"hdfs://localhost:8020/accumulo/tables/2a/default_tablet/F0000072.rf\",\"startRow\":\"\",\"endRow\":\"\"}"]}> but was: <{"txid":"FATE:USER:0000000000000002","selAll":true,"files":["{\"path\":\"hdfs://localhost:8020/accumulo/tables/2a/default_tablet/F0000070.rf\",\"startRow\":\"\",\"endRow\":\"\"}","{\"path\":\"hdfs://localhost:8020/accumulo/tables/2a/default_tablet/F0000071.rf\",\"startRow\":\"\",\"endRow\":\"\"}","{\"path\":\"hdfs://localhost:8020/accumulo/tables/2a/default_tablet/F0000072.rf\",\"startRow\":\"\",\"endRow\":\"\"}"]}>
```

**Comment by kevinrr888** [comment]
> @kevinrr888 - The latest IT run is failing (stacktrace below). I think it might be due to this merge. Just an FYI...
> 
> ```
> org.opentest4j.AssertionFailedError: Test json should be identical to actual metadata at this point ==> expected: 
> <{"fateId":"FATE:USER:0000000000000002","selAll":true,"files":["
> {\"path\":\"hdfs://localhost:8020/accumulo/tables/2a/default_tablet/F0000070.rf\",\"startRow\":\"\",\"endRow\":\"\"}","
> {\"path\":\"hdfs://localhost:8020/accumulo/tables/2a/default_tablet/F0000071.rf\",\"startRow\":\"\",\"endRow\":\"\"}","
> {\"path\":\"hdfs://localhost:8020/accumulo/tables/2a/default_tablet/F0000072.rf\",\"startRow\":\"\",\"endRow\":\"\"}"]}> 
> but was: <{"txid":"FATE:USER:0000000000000002","selAll":true,"files":["
> {\"path\":\"hdfs://localhost:8020/accumulo/tables/2a/default_tablet/F0000070.rf\",\"startRow\":\"\",\"endRow\":\"\"}","
> {\"path\":\"hdfs://localhost:8020/accumulo/tables/2a/default_tablet/F0000071.rf\",\"startRow\":\"\",\"endRow\":\"\"}","
> {\"path\":\"hdfs://localhost:8020/accumulo/tables/2a/default_tablet/F0000072.rf\",\"startRow\":\"\",\"endRow\":\"\"}"]}>
> ```

Thanks for pointing this out. I created https://github.com/apache/accumulo/pull/4265

**Pull request merged**